packages/cache-map/src/cachemap.js

```diff
@@ -0,0 +1,322 @@
+ * {@link makeCacheMapKit}.
+ *
+ * @template K
+ * @template V
+ * @typedef {Pick<Map<K, V>, Exclude<keyof WeakMap<WeakKey, *>, 'set'>> & {set: (key: K, value: V) => WeakMapAPI<K, V>}} WeakMapAPI
+ */
+
+/**
+ * @template K
+ * @template V
+ * @typedef {WeakMapAPI<K, V> & {size?: (0 | 1)}} SingleEntryMap
+ */
+
+/**
+ * A cell of a doubly-linked ring (circular list) for a cache map.
+ * Instances are not frozen, and so should be closely encapsulated.
+ *
+ * @template K
+ * @template V
+ * @typedef {object} CacheMapCell
+ * @property {number} id for debugging
+ * @property {CacheMapCell<K, V>} next
+ * @property {CacheMapCell<K, V>} prev
+ * @property {SingleEntryMap<K, V>} data
+ */
+
+/**
+ * @template K
+ * @template V
+ * @param {CacheMapCell<K, V>} prev
+ * @param {number} id
+ * @param {WeakMapAPI<K, V>} data
+ * @returns {CacheMapCell<K, V>}
+ */
+const appendNewCell = (prev, id, data) => {
+  const next = prev?.next;
+  const cell = { id, next, prev, data };
+  if (prev) prev.next = cell;
+  if (next) next.prev = cell;
+  return cell;
+};
+
+/**
+ * @template K
+ * @template V
+ * @param {CacheMapCell<K, V>} cell
+ * @param {CacheMapCell<K, V>} prev
+ * @param {CacheMapCell<K, V>} [next]
+ */
+const moveCellAfter = (cell, prev, next = prev.next) => {
+  if (cell === prev || cell === next) return; // already in position
+
+  // Splice out cell.
+  const { prev: oldPrev, next: oldNext } = cell;
+  oldPrev.next = oldNext;
+  oldNext.prev = oldPrev;
+
+  // Splice in cell after prev.
+  cell.prev = prev;
+  cell.next = next;
+  prev.next = cell;
+  next.prev = cell;
+};
+
+/**
+ * Clear out a cell to prepare it for future use. Its map is preserved when
+ * possible, but must instead be replaced if the associated key is not known.
+ *
+ * @template K
+ * @template V
+ * @param {CacheMapCell<K, V>} cell
+ * @param {K | UNKNOWN_KEY} oldKey
+ * @param {() => WeakMapAPI<K, V>} [makeMap] required when the key is unknown
+ */
+const resetCell = (cell, oldKey, makeMap) => {
+  if (oldKey === UNKNOWN_KEY) {
+    const { size } = cell.data;
+    if (size === undefined) {
+      // WeakMap instances must be replaced when the key is unknown.
+      if (!makeMap) {
+        throw Error('internal: makeMap is required with UNKNOWN_KEY');
+      }
+      cell.data = makeMap();
+      return;
+    }
+    // Map instances can always be cleared for reuse.
+    if (size === 0) return;
+    if (size !== 1) {
+      throw Error('internal: Unexpected non-singular cell data entry count');
+    }
+    const cellData = /** @type {Map<K, V>} */ (cell.data);
+    oldKey = /** @type {K} */ (cellData.keys().next().value);
+  }
+  cell.data.delete(oldKey);
+};
+
+const zeroMetrics = freeze({
+  totalQueryCount: 0,
+  totalHitCount: 0,
+  // TODO?
+  // * method-specific counts
+  // * liveTouchStats/evictedTouchStats { count, sum, mean, min, max }
+  //   * p50/p90/p95/p99 via Ben-Haim/Tom-Tov streaming histograms
```

Comment:
A former colleague referred to count/sum/mean/min/max/p50/p90/p95/p99/p99.9 as the "big ten" metrics[^1].

[^1]: I believe [this](https://www.oracle.com/cloud/streaming/faq/#monitoring) is the result of that influence.

Reply:
They’re pretty ubiquitous, to be sure.